lark/load_grammar.py

```diff
@@ -648,17 +649,67 @@ def compile(self, start):
         return terminals, compiled_rules, self.ignore
 
 
+class FromPackageLoader(object):
+    """
+    Provides a simple way of creating custom import loaders that load from packages via ``pkgutil.get_data`` instead of using `open`.
+    This allows them to be compatible even from within zip files.
+    
+    Relative imports are handled, so you can just freely use them.
+    
+    pkg_name: The name of the package. You can probably provide `__name__` most of the time
+    search_paths: All the path that will be search on absolute imports.
+    """
+    def __init__(self, pkg_name, search_paths=("", )):
+        self.pkg_name = pkg_name
+        self.search_paths = search_paths
+        
+    def __repr__(self):
+        return "%s(%r, %r)" % (type(self).__name__, self.pkg_name, self.search_paths)
+
+    def __call__(self, base_paths, grammar_path):
+        if len(base_paths) == 0:
+            to_try = self.search_paths
+        else:
+            assert len(base_paths) == 1
+            if not base_paths[0].startswith('<%s:' % (self.pkg_name,)):
```

Comment:
What is this?

Reply:
This is to recognize relative imports from within a package. When a grammar is loaded by `FromPackageLoader`, it gets a path in the form `<{pkg_name}:/{full_path}>` Where `full_path` is the argument that was passed to `pkgutil.get_data`. Here we unparse it to get the relative path that was specified in importing (allowing relative imports). I will write a comment for this.

Reply:
Why are we accepting this repr-like string? Just so we can support `__name__`?

Reply:
No, because there might not be a meaningful way of expressing the location of the grammar file as a normal file path the `open` accepts. This is similar to `<string>`, just that it also carries information for imports inside of it instead of implied/inferred at a latter stage. This is a format I designed to just be used internally by the `FromPackageLoader` and also be readable by the end user. Do you have an alternative suggestion?

Reply:
Well, if you already know the package name.. Why not just accept a regular path within the package?

Reply:
Because that could cause problems when there are multiple packages from which we are loading. And it is not like these paths would actually be useful for anything else.

Reply:
Maybe I missing something, but this all seems a little messier than it has to be.

But if you insist on this hack, parse it with a regexp. It shouldn't take more than two lines.

And document it well.

Reply:
If you have an alternative suggestion that doesn't run into problems when there are multiple files with the same name in different places, I am open to ideas.

Reply:
So, I'm trying to think of the actual use cases. Help me understand.

The first use case, is when trying to load the main grammar from a package. But there is a method for it, and it's only one grammar, one package. No difficulty.

The second use case, is when trying to import a grammar from within a grammar. In that case, we currently don't allow to specify the package anyway. We only allow a dotted path.

I can't think of other use cases.

So, in what realistic situation will this feature be used?

Reply:
Maybe one can create a package with grammar library with lots of different grammar files. Then you can add that package to your own module where you also have a somewhat complex grammar split over multiple files. (Or a plugin system). If you try to use relative imports, we have to prevent the grammars from one package to find the grammars in the other package.

Under the old file-based system, this wasn't a problem since we always specified an absolute file path that clearly defines which grammar file we are talking about.

But now we can't easily do that. If we just use `full_path` there might be ambiguity of both packages use the `grammars` folder. If we try to use an absolute path, this will break when we are inside of a zip file. So we somehow have to encode both the package and the path inside of the package, unless we don't want to support those usecases. And I don't see any real harm done by this.

Reply:
You are describing a situation that no one currently has.

Also, I still don't understand who will pass the `<package:path>` string, and where it will pass.

If both packages have a `grammars` folder in the search paths, that is still not a collision. Only if they have the same `whatever.lark` file it will collide, no? And then it will choose whatever is first on the list.

The harm is code bloat. Death by a million cuts. Every little feature we add, is something we will have to support. And one line after another end up being hundreds of lines, that force us to do things in a certain way, because we overengineered when we didn't have to.

Again, maybe if you can give a specific use case that it will solve, it will help me understand. If you can't explain it, show me the code from the user's perspective.

Reply:
> Also, I still don't understand who will pass the <package:path> string, and where it will pass.
This is the end of the function. I added a few more comments. So it is just the `FromPackageLoader` talking to itself.

Reply:
I see. If this is an internal structure, and not something that is user-facing, then please let's use a dedicated class for it.

Reply:
Well, it is also the name of the grammar to be used in error reporting (the source argument to `GrammarLoader.load_grammar`) The benefit of just using str is that we don't need to make any changes in `load_grammar`.

Reply:
That's what `__str__` is for. And if we need to add a `str()` call somewhere, it will be well worth it.

Reply:
Well, I pushed another commit. I would call this more code bloat than before, but if you prefer this, ok.

Reply:
To be fair, this whole zipapp feature was a code bloat :)